// @cliqz/adblocker/dist/cjs/src/engine/metadata/categories.js
"use strict";
/*!
 * Copyright (c) 2017-present Cliqz GmbH. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.createMap = exports.deserialize = exports.serialize = exports.getSerializedSize = exports.getKey = exports.isValid = void 0;
/* tslint:disable variable-name */
const map_1 = require("../map");
const data_view_1 = require("../../data-view");
const utils_1 = require("../../utils");
function isValid(category) {
    if (category === null) {
        return false;
    }
    if (typeof category !== 'object') {
        return false;
    }
    const { key, name, color, description } = category;
    if (typeof key !== 'string') {
        return false;
    }
    if (typeof name !== 'string') {
        return false;
    }
    if (typeof color !== 'string') {
        return false;
    }
    if (typeof description !== 'string') {
        return false;
    }
    return true;
}
exports.isValid = isValid;
function getKey(category) {
    return (0, utils_1.fastHash)(category.key);
}
exports.getKey = getKey;
function getSerializedSize(category) {
    return ((0, data_view_1.sizeOfUTF8)(category.key) +
        (0, data_view_1.sizeOfUTF8)(category.name) +
        (0, data_view_1.sizeOfUTF8)(category.color) +
        (0, data_view_1.sizeOfUTF8)(category.description));
}
exports.getSerializedSize = getSerializedSize;
function serialize(category, view) {
    view.pushUTF8(category.key);
    view.pushUTF8(category.name);
    view.pushUTF8(category.color);
    view.pushUTF8(category.description);
}
exports.serialize = serialize;
function deserialize(view) {
    return {
        key: view.getUTF8(),
        name: view.getUTF8(),
        color: view.getUTF8(),
        description: view.getUTF8(),
    };
}
exports.deserialize = deserialize;
function createMap(categories) {
    return new map_1.CompactMap({
        getSerializedSize,
        getKeys: (category) => [getKey(category)],
        serialize,
        deserialize,
        values: categories,
    });
}
exports.createMap = createMap;
